### ISSUE TITLE
04. PR3: TraceMemoryConsistencyCriterion と factory を実装する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#435]
# 04. PR3: TraceMemoryConsistencyCriterion と factory を実装する

## 目的

既存 coarse criterion を wrap し、trace memory consistency loss を加算する criterion class を実装する。

この issue では class と factory を実装するが、`train.py` への接続は行わない。接続は PR4 で扱う。

## 依存 issue

- `01_issue_add_trace_memory_module_contract.md`
- `02_issue_implement_sparse_trace_pick_memory_bank.md`
- `03_issue_add_projection_and_gaussian_target_helpers.md`

## 対象ファイル

```text
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/trace_memory_consistency.py
```

必要 import:

```python
import torch
from .output import get_pick_logits
from .config import CoarseTraceMemoryConsistencyCfg
```

## 追加する class

```python
class TraceMemoryConsistencyCriterion:
    def __init__(
        self,
        *,
        base_criterion,
        cfg: CoarseTraceMemoryConsistencyCfg,
        fb_sigma_ms: float,
        memory_bank: SparseTracePickMemoryBank | None = None,
    ) -> None:
        ...
```

`cfg.enabled` が false の場合は、この class を直接作っても動くようにする。ただし通常は factory が base criterion をそのまま返す。

`fb_sigma_ms` は `cfg.sigma_ms is None` のときの fallback。

## 追加する factory

```python
def maybe_wrap_trace_memory_consistency(
    *,
    base_criterion,
    cfg: CoarseTraceMemoryConsistencyCfg,
    fb_sigma_ms: float,
    memory_bank: SparseTracePickMemoryBank | None = None,
):
    ...
```

挙動:

```text
cfg.enabled == False or cfg.weight == 0 and cfg.update_min_pmax == 0?
  原則 base_criterion を返すのではなく、enabled false のときだけ base を返す。

cfg.enabled == True:
  TraceMemoryConsistencyCriterion(...) を返す。
```

`weight=0` でも enabled true なら wrap する。memory update の smoke / diagnostics に使えるため。

## `__call__` の流れ

```python
def __call__(self, pred: object, target: torch.Tensor, batch: dict) -> torch.Tensor:
    base_loss = self.base_criterion(pred, target, batch)
    pick_logits = get_pick_logits(pred)

    cons_loss = self.compute_consistency_loss(pick_logits, batch)
    weight = self.effective_weight()
    total = base_loss + weight * cons_loss

    self.update_memory(pick_logits.detach(), batch)
    self.batch_step += 1
    return total
```

重要: **loss 計算後に memory update** する。同じ batch の prediction をその場で target にしない。

## effective weight

```python
def effective_weight(self) -> float:
    if self.batch_step < cfg.warmup_batches:
        return 0.0
    if cfg.ramp_batches <= 0:
        return cfg.weight
    r = min(1.0, max(0.0, (self.batch_step - cfg.warmup_batches) / cfg.ramp_batches))
    return cfg.weight * r
```

`batch_step` は mini-batch 数でよい。optimizer step 数ではない。

## current prediction extraction

`pick_logits` shape は `(B,1,H,W)`。

```python
prob = torch.softmax(pick_logits.detach().float().squeeze(1), dim=-1)
pmax, pick_view = prob.max(dim=-1)  # (B,H)
```

memory update 用の `pick_view` は detached でよい。

consistency loss 用には gradient が必要な current distribution を使う。

```python
log_p = torch.log_softmax(pick_logits.float().squeeze(1), dim=-1)  # (B,H,W)
```

## consistency target selection

batch を flatten して row 単位で処理する。

```text
sample_index: 0..B-1 repeated H times
row_index:    0..H-1 tiled B times
```

candidate mask:

```text
trace_valid == true
anchor_raw_indices_view >= 0
memory lookup hit
memory count >= cfg.min_count
memory pmax >= cfg.target_min_pmax
age <= cfg.max_age_batches        # cfg.max_age_batches is not None のとき
memory_view in [0, W - 1]
label_guard を満たす              # cfg.label_guard_samples is not None のとき
```

label guard:

```python
labeled = fb_idx_view > 0
ok = (~labeled) | (abs(memory_view - fb_idx_view.float()) <= cfg.label_guard_samples)
```

`fb_idx_view <= 0` は unlabeled/ignored とみなす。既存 coarse label contract と合わせる。

## consistency loss

推奨は Gaussian target cross entropy。

```python
q = build_trace_memory_gaussian_targets(...)
loss_per_trace = -(q * log_p_flat[used_rows]).sum(dim=-1)
cons_loss = loss_per_trace.mean()
```

`cfg.sigma_ms is None` の場合は `fb_sigma_ms` を使う。

used row が 0 の場合:

```python
return pick_logits.sum() * 0.0
```

これにより `total_loss` は常に tensor graph と整合する。

## memory update

`update_memory()` は current prediction を raw coordinate に戻して memory bank を更新する。

```text
current_raw = (current_view_pick + start - trace_tshift_view) / factor
slot = current primary/secondary
```

update mask:

```text
trace_valid == true
anchor_raw_indices_view >= 0
current_raw finite
pmax >= cfg.update_min_pmax
```

`SparseTracePickMemoryBank.update(...)` に flat tensors を渡す。

## metrics

PR5 の logging 接続に備えて、criterion 側にも軽い metrics API を入れる。

```python
def consume_stats(self) -> dict[str, float]:
    ...
```

最低限返す値:

```text
trace_memory/loss_sum
trace_memory/loss_count
trace_memory/used_rows
trace_memory/candidate_rows
trace_memory/low_pmax_reject_rows
trace_memory/min_count_reject_rows
trace_memory/age_reject_rows
trace_memory/label_guard_reject_rows
trace_memory/effective_weight
```

`memory_bank.consume_stats()` の結果も prefix を維持して merge してよい。

PR3 では train loop から呼ばれないが、unit test で確認できる状態にする。

## `__all__`

以下を追加する。

```python
'TraceMemoryConsistencyCriterion',
'maybe_wrap_trace_memory_consistency',
```

## 非目標

この issue では以下を実装しない。

- `train.py` での factory 呼び出し
- train skeleton からの metrics logging
- checkpoint への memory 保存
- distributed training 対応
- site54 overlay

## 受け入れ条件

- memory が空のとき、criterion は base loss と同じ値を返し、呼び出し後に memory が update される。
- 2 回目以降、policy と gate を満たす memory target がある場合に consistency loss が正になる。
- `primary_cross` は same-primary target を使わない。
- `secondary_cross` は same-primary かつ cross-secondary target のみ使う。
- `warmup_batches` / `ramp_batches` により effective weight が仕様通りになる。
- `label_guard_samples` が supervised label と矛盾する target を捨てる。
- used row が 0 でも backward 可能な finite tensor を返す。

## 確認コマンド

```bash
python -m py_compile \
  packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/trace_memory_consistency.py
```
